### PR TITLE
secure_services: Lock scheduler before calling secure services.

### PR DESCRIPTION
**Pull request by oyvindronningstad**
call k_sched_lock before and k_sched_unlock after.

Rename the services from spm_foo() to spm_foo_nsc() and call
spm_foo_nsc() from spm_foo() along with k_sched_*()

Ref: NCSIDB-108

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
If https://github.com/zephyrproject-rtos/zephyr/pull/26414 goes through, this PR should be changed to accommodate

**Review by ioannisg — DISMISSED** [review]
I would believe that the scheduler lock needs to happen before the secure call (at least this is how I described the proposed solution in the respective ticket). Otherwise we do not completely get rid of race conditions. [Might be missing something so let's discuss.]

**Comment by oyvindronningstad** [comment]
> I would believe that the scheduler lock needs to happen before the secure call (at least this is how I described the proposed solution in the respective ticket). Otherwise we do not completely get rid of race conditions. [Might be missing something so let's discuss.]

It does happen before the secure call. The _nsc functions are the entry functions.

**Comment by ioannisg** [comment]
Ok, thanks, I saw it now. So the patch should be functionally fine, however I think it would be better to embed the cooperative scheduling enforcement in the way the Secure calls are defined. So you would not need to duplicate the kernel locking in every secure service call. And that patch could actually be sent upstream directly!

**Comment by ioannisg** [comment]
>Ok, thanks, I saw it now. So the patch should be functionally fine, however I think it would be better to embed the cooperative scheduling enforcement in the way the Secure calls are defined. So you would not need to duplicate the kernel locking in every secure service call. And that patch could actually be sent upstream directly!

This work was merged upstream in https://github.com/zephyrproject-rtos/zephyr/pull/27831/ and should be part of v2.4.0. I believe @oyvindronningstad has updated the current PR to take the upstream changes into account.

**Review by lemrey — APPROVED** [review]
Thanks, this fixed the issue I was seeing when calling `sys_rand32_get`.
I wonder if it would be better to infix `_ncs_` instead e.g. `spm_ncs_foo`, no strong opinion though.

nit: can you remove the punctuation from the commit title

File: subsys/spm/secure_services_ns.c [hunk 1]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
Comment (hakonfam) [inline subsys/spm/secure_services_ns.c:2]:
```suggestion
 * Copyright (c) 2020 Nordic Semiconductor ASA
```

File: subsys/spm/secure_services_ns.c [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <zephyr.h>
+#include <errno.h>
+#include <../arch/arm/include/aarch32/cortex_m/tz_ns.h>
+#include <fw_info.h>
+
+void spm_request_system_reboot_nsc(void);
```
Comment (hakonfam) [inline subsys/spm/secure_services_ns.c:11]: Why are these needed? Would it be possible to get rid of them by adding a new layer of macros?

File: subsys/spm/secure_services_ns.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+int spm_request_random_number_nsc(uint8_t *output, size_t len, size_t *olen);
+int spm_request_read_nsc(void *destination, uint32_t addr, size_t len);
+int spm_firmware_info_nsc(uint32_t fw_address, struct fw_info *info);
+int spm_prevalidate_b1_upgrade_nsc(uint32_t dst_addr, uint32_t src_addr);
+void spm_busy_wait_nsc(uint32_t busy_wait_us);
+
+TZ_THREAD_SAFE_NONSECURE_ENTRY_FUNC(spm_request_system_reboot,
```
Comment (hakonfam) [inline subsys/spm/secure_services_ns.c:18]: I guess that the commit message was composed before this functionality was available in zephyr, so you should check that it is up to date. 

```
call k_sched_lock before and k_sched_unlock after.

Rename the services from spm_foo() to spm_foo_nsc() and call
spm_foo_nsc() from spm_foo() along with k_sched_*()
```

There is no mention of `k_sched` in this PR since it is all handled in the TZ macros.

File: include/secure_services.h [hunk 4]
```diff
@@ -22,6 +22,7 @@
 
 #include <stddef.h>
 #include <zephyr/types.h>
+#include <zephyr.h>
```
Comment (hakonfam) [inline include/secure_services.h:25]: Are the changes to this file needed?

File: subsys/spm/Kconfig [hunk 5]
```diff
@@ -111,6 +111,13 @@ config SPM_SERVICE_PREVALIDATE
 	  secure service is needed for the app to access the prevalidation
 	  function.
 
+config SPM_SERVICE_BUSY_WAIT
+	bool "Busy wait in secure mode (debug function)"
+	default y
+	help
```
Comment (hakonfam) [inline subsys/spm/Kconfig:117]: Could it depend on ZTEST so that its not visible for customers when running menuconfig on a normal project?

File: tests/subsys/spm/thread_swap/testcase.yaml [hunk 6]
```diff
@@ -0,0 +1,4 @@
+tests:
+  spm.secure_services.thread_swap:
+    platform_allow: nrf9160dk_nrf9160ns nrf5340pdk_nrf5340_cpuappns
```
Comment (hakonfam) [inline tests/subsys/spm/thread_swap/testcase.yaml:3]:
```suggestion
    platform_allow: nrf9160dk_nrf9160ns nrf5340pdk_nrf5340_cpuappns nrf5340dk_nrf5340_cpuappns
```

Comment (hakonfam) [reply tests/subsys/spm/thread_swap/testcase.yaml:3]: Not really sure if this is needed, just check what is in the other tests.

File: tests/subsys/spm/thread_swap/src/main.c [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+static void work_func(struct k_work *work)
+{
+	work_done = true;
+	spm_busy_wait(20000);
+}
+
+void test_spm_service_thread_swap1(void)
+{
+	int err;
```
Comment (hakonfam) [inline tests/subsys/spm/thread_swap/src/main.c:22]: A comment on what is being tested would be good, currently you have to read the commit message to see a description of what this test does and why its needed.

File: subsys/CMakeLists.txt [hunk 8]
```diff
@@ -21,6 +21,7 @@ add_subdirectory_ifdef(CONFIG_PROFILER		profiler)
 add_subdirectory_ifdef(CONFIG_DFU_TARGET		dfu)
 add_subdirectory_ifdef(CONFIG_PCD 		pcd)
 add_subdirectory_ifdef(CONFIG_IS_SPM		spm)
+add_subdirectory_ifdef(CONFIG_SPM		spm)
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply subsys/CMakeLists.txt:24]: agreed this needs explanation

File: subsys/spm/Kconfig (hunk 5 above)
Comment (ioannisg) [reply subsys/spm/Kconfig:117]: Yes this is an option. (or TEST)
How about samples, where we would like to have this set by default as well?

File: subsys/spm/secure_services_ns.c [hunk 9]
```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+#include <zephyr.h>
+#include <errno.h>
+#include <../arch/arm/include/aarch32/cortex_m/tz_ns.h>
```
[the remark this answers is not in the document]
Comment (ioannisg) [reply subsys/spm/secure_services_ns.c:8]: I think it is not by default, but if it is, yeah, this should be simplified.

File: subsys/spm/secure_services_ns.c (hunk 3 above)
Comment (ioannisg) [reply subsys/spm/secure_services_ns.c:18]: and by the way, you should elaborate here why THREAD_SAFE is needed (maybe some simple inline comments above the definitions)

**Review by nordic-hani — APPROVED** [review]

File: subsys/spm/secure_services_ns.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/spm/secure_services_ns.c:8]: This is ... surprisingly complicated. I want to leave it like this here, and start a process upstream independently.

File: subsys/spm/Kconfig (hunk 5 above)
Comment (oyvindronningstad) [reply subsys/spm/Kconfig:117]: Sadly, this config (in SPM) cannot depend on a config in the app.

**Comment by tejlmand** [comment]
@oyvindronningstad retriggered CI by forced push: 22d3fb81a24121f42793dba6fa84cf54fe3688de --> 22d3fb81a24121f42793dba6fa84cf54fe3688de

File: subsys/spm/Kconfig (hunk 5 above)
Comment (hakonfam) [reply subsys/spm/Kconfig:117]: But what if we modify the build system to pass "CONFIG_ZTEST" to any child image? In my opinion that is a reasonable feature. If you agree we can create a task and add it to the backlog.

**Comment by hakonfam** [comment]
Rebased on latest master and force pushed from 22d3fb8 to 10e6282 to try and fix CI.

**Comment by hakonfam** [comment]
Hmm, look like there's legit failures in CI

```
/opt/gnuarmemb/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: zephyr/libzephyr.a(secure_services_ns.c.obj): in function `spm_request_system_reboot':
/jenkins_workspace/workspace/nrfconnect-boot-fw-update_master/test_objects/nrf/subsys/spm/secure_services_ns.c:18: undefined reference to `spm_request_system_reboot_nsc'
/opt/gnuarmemb/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: zephyr/libzephyr.a(secure_services_ns.c.obj): in function `spm_prevalidate_b1_upgrade':
/jenkins_workspace/workspace/nrfconnect-boot-fw-update_master/test_objects/nrf/subsys/spm/secure_services_ns.c:33: undefined reference to `spm_prevalidate_b1_upgrade_nsc'
collect2: error: ld returned 1 exit status
```

File: subsys/spm/Kconfig (hunk 5 above)
Comment (oyvindronningstad) [reply subsys/spm/Kconfig:117]: I spent a bit of time thinking about this yesterday. I had a similar thought. We have the option of passing `-Dspm_CONFIG_TEST=y`-style Kconfig overrides from app to spm, but it's quite blunt, and means that those configs will be ignored if set with e.g. `ninja spm_menuconfig`.

I also considered just doing it with the SPM_SERVICE configs since they anyway have particular requirements about being in sync across app and spm. Doing this would solve another problem as well, which I fix in https://github.com/nrfconnect/sdk-nrf/pull/2710, but I like this solution better.

File: subsys/spm/secure_services_ns.c (hunk 9 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/spm/secure_services_ns.c]: https://github.com/zephyrproject-rtos/zephyr/issues/29160

File: subsys/CMakeLists.txt (hunk 8 above)
[the remark this answers is not in the document]
Comment (oyvindronningstad) [reply subsys/CMakeLists.txt]: Removed.

File: subsys/spm/secure_services_ns.c (hunk 2 above)
Comment (oyvindronningstad) [reply subsys/spm/secure_services_ns.c]: Done

File: subsys/spm/secure_services_ns.c (hunk 3 above)
Comment (oyvindronningstad) [reply subsys/spm/secure_services_ns.c]: Reworded the commit message. I didn't add anything about thread safety here, but I added docs elsewhere.

File: subsys/spm/Kconfig (hunk 5 above)
Comment (oyvindronningstad) [reply subsys/spm/Kconfig:125]: The SPM_SERVICE_* configs are now passed from app to spm

File: tests/subsys/spm/thread_swap/src/main.c (hunk 7 above)
Comment (oyvindronningstad) [reply tests/subsys/spm/thread_swap/src/main.c:26]: Done, take a look

**Comment by oyvindronningstad** [comment]
> 
> 
> Hmm, look like there's legit failures in CI
> 
> ```
> /opt/gnuarmemb/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: zephyr/libzephyr.a(secure_services_ns.c.obj): in function `spm_request_system_reboot':
> /jenkins_workspace/workspace/nrfconnect-boot-fw-update_master/test_objects/nrf/subsys/spm/secure_services_ns.c:18: undefined reference to `spm_request_system_reboot_nsc'
> /opt/gnuarmemb/gcc-arm-none-eabi-9-2019-q4-major/bin/../lib/gcc/arm-none-eabi/9.2.1/../../../../arm-none-eabi/bin/ld: zephyr/libzephyr.a(secure_services_ns.c.obj): in function `spm_prevalidate_b1_upgrade':
> /jenkins_workspace/workspace/nrfconnect-boot-fw-update_master/test_objects/nrf/subsys/spm/secure_services_ns.c:33: undefined reference to `spm_prevalidate_b1_upgrade_nsc'
> collect2: error: ld returned 1 exit status
> ```

I couldn't reproduce this locally with samples, so the test might be misconfigured. Will look more into later.

File: samples/CMakeLists.txt [hunk 10]
```diff
@@ -5,6 +5,21 @@
...
+  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
+
+  set(regex "^CONFIG_SPM_SERVICE.+")
+
+  list(FILTER VARIABLES        INCLUDE REGEX ${regex})
+  list(FILTER VARIABLES_CACHED INCLUDE REGEX ${regex})
+
+  foreach(var_name ${VARIABLES} ${VARIABLES_CACHED})
+    set(spm_${var_name} ${${var_name}})
+    print(${var_name})
```
Comment (hakonfam) [inline samples/CMakeLists.txt:18]: intended?

Comment (oyvindronningstad) [reply samples/CMakeLists.txt:18]: Nope :D

File: include/secure_services.h [hunk 11]
```diff
@@ -22,12 +22,35 @@
...
+ *              to be named the same, but with the suffix '_nsc'. E.g. the
+ *              wrapper function foo() wraps the secure service foo_nsc().
+ * @param ...   The arguments of the secure service and the wrapper function, as
+ *              they would appear in a function signature, i.e. type and name.
+ */
+#define NRF_NSC(ret, name, ...) \
+ret name ## _nsc(__VA_ARGS__); \
+TZ_THREAD_SAFE_NONSECURE_ENTRY_FUNC(name, ret, name ## _nsc, __VA_ARGS__)
+
+
```
Comment (tejlmand) [inline include/secure_services.h:53]: Why double empty line ?
```suggestion
```

File: include/secure_services.h [hunk 12]
```diff
@@ -91,6 +114,16 @@ int spm_firmware_info(uint32_t fw_address, struct fw_info *info);
...
+/** Busy wait in secure mode (debug function)
+ *
+ * This function is for writing tests that require the execution to be in
+ * secure mode
+ *
+ * @param[in]  busy_wait_us  The number of microseconds to wait for.
+ */
+void spm_busy_wait(uint32_t busy_wait_us);
+
+
```
Comment (tejlmand) [inline include/secure_services.h:126]: Why double empty line ?
```suggestion
```

File: samples/CMakeLists.txt [hunk 13]
```diff
@@ -5,6 +5,20 @@
 #
 
 if (CONFIG_SPM)
+  get_cmake_property(VARIABLES              VARIABLES)
+  get_cmake_property(VARIABLES_CACHED CACHE_VARIABLES)
+
+  set(regex "^CONFIG_SPM_SERVICE.+")
+
+  list(FILTER VARIABLES        INCLUDE REGEX ${regex})
+  list(FILTER VARIABLES_CACHED INCLUDE REGEX ${regex})
```
Comment (tejlmand) [inline samples/CMakeLists.txt:14]: This is ok for now, but `VARIABLES` is a very long list.
Is this really the only way to do this ?

File: samples/CMakeLists.txt [hunk 14]
```diff
@@ -5,6 +5,20 @@
...
+
+  foreach(var_name ${VARIABLES} ${VARIABLES_CACHED})
+    set(spm_${var_name} ${${var_name}})
```
Comment (tejlmand) [inline samples/CMakeLists.txt:17]: So, if user did: `-Dspm_CONFIG_SPM_SERVICE<something>=<value>` that will be lost and overwritten with the value of `CONFIG_SPM_SERVICE<something>`.

Sounds like a flaw.

File: tests/subsys/spm/thread_swap/src/main.c [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+	/* Call into the secure service which will be interrupted. If the
+	 * scheduler is locked before entering secure services, it will prevent
+	 * work_func from being run until after the below function returns, but
+	 * this is valid behavior.
+	 */
+	spm_busy_wait(20000);
+	zassert_true(work_done, "work_func should have interrupted wait.");
+}
+
```
Comment (tejlmand) [inline tests/subsys/spm/thread_swap/src/main.c:46]:
```suggestion
```

**Review by tejlmand — COMMENTED** [review]
Some small nits, and one possible flaw but not necessarily blocking for merging.

File: samples/CMakeLists.txt (hunk 13 above)
Comment (oyvindronningstad) [reply samples/CMakeLists.txt:14]: It's the only way I could think of where we won't have to remember to edit this code every time a new secure service is added. I totally agree it's not ideal.

Comment (tejlmand) [reply samples/CMakeLists.txt:14]: yep, let's look into an improvement later.
Could you create an issue and assign to me ?

File: samples/CMakeLists.txt (hunk 14 above)
Comment (oyvindronningstad) [reply samples/CMakeLists.txt:17]: In the case of these configs, any mismatch in configuration between SPM and app is an error, so in this case, it's "desired". Ideally, there should be a way to prevent setting the config in SPM, but I opted to just mention it in the help text.

File: tests/subsys/spm/thread_swap/testcase.yaml (hunk 6 above)
Comment (oyvindronningstad) [reply tests/subsys/spm/thread_swap/testcase.yaml:3]: Looks like nrf5340dk_nrf5340_cpuappns is not used in any testcase.yaml or sample.yaml files yet.

File: samples/CMakeLists.txt (hunk 13 above)
Comment (oyvindronningstad) [reply samples/CMakeLists.txt:14]: Done

**Comment by oyvindronningstad** [comment]
> Thanks, this fixed the issue I was seeing when calling `sys_rand32_get`.
> I wonder if it would be better to infix `_ncs_` instead e.g. `spm_ncs_foo`, no strong opinion though.

Thanks. This is now used via concatenation though, so infix is not really a convenient alternative.

> nit: can you remove the punctuation from the commit title

Thanks, done

File: subsys/spm/secure_services.c [hunk 16]
```diff
@@ -66,7 +68,7 @@ struct read_range {
 
 
 __TZ_NONSECURE_ENTRY_FUNC
-int spm_request_read(void *destination, uint32_t addr, size_t len)
+int spm_request_read_nsc(void *destination, uint32_t addr, size_t len)
```
Comment (frkv) [inline subsys/spm/secure_services.c:71]: I've recommended renaming this from "nsc" to "nse" for **N**on-**S**ecure **E**ntry to conform to the macro used

Comment (oyvindronningstad) [reply subsys/spm/secure_services.c]: Thanks, done

**Review by frkv — APPROVED** [review]
LGTM

File: samples/CMakeLists.txt (hunk 14 above)
Comment (tejlmand) [reply samples/CMakeLists.txt:17]: accepted.
Probably we should warn the user then, in case he has already defined `spm_CONFIG_SPM_SERVICE<something>` before overwriting the value.

But that  can be done as enhancement.

**Review by tejlmand — APPROVED** [review]

**Review by hakonfam — APPROVED** [review]

**Review by shantha-14 — APPROVED** [review]
